KAFKA-15358: Added QueuedReplicaToDirAssignments metric

mwesterby opened the pull request:
Adds a new metric, QueuedReplicaToDirAssignments, which, when queried, will return the total number of replica to directory assignments that are currently queued in the AssignmentsManager.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

soarez reviewed (approved):
LGTM

rondagostino commented on server/src/main/java/org/apache/kafka/server/AssignmentsManager.java:
```diff
@@ -240,31 +269,53 @@ public void run() throws Exception {
 
+                log.warn("Re-queueing assignments: {}", failed);
                 if (!failed.isEmpty()) {
-                    log.warn("Re-queueing assignments: {}", failed);
```
nit: seems odd to log a warning even for the case where the set is empty.

rondagostino reviewed (approved):
LGTM -- thanks for the PR!  Just one minor comment about a log level change.  

`There are 6 new tests failing, 4 existing failing` -- nothing fails under multiple JDKs, and all failures seem unrelated.

Will wait for @cmccabe to take a quick look before merging since he last touched this PR.

rondagostino commented:
Can we get the conflicts resolved?

rondagostino commented:
@mwesterby Can you resolve the conflicts here?

mwesterby committed:
KAFKA-15358: Added QueuedReplicaToDirAssignments metric

Adds a new metric, QueuedReplicaToDirAssignments, which, when queried,
will return the total number of replica to directory assignments that
are currently queued in the AssignmentsManager.

soarez commented on server/src/main/java/org/apache/kafka/server/AssignmentsManager.java:
```diff
@@ -93,10 +100,24 @@ public AssignmentsManager(Time time,
+        this.metricsGroup.newGauge(QUEUE_REPLICA_TO_DIR_ASSIGNMENTS_METRIC_NAME, new Gauge<Integer>() {
+            @Override
+            public Integer value() {
+                return getMapSize(inflight) + getMapSize(pending);
```
The map references are copied into `getMapSize` which avoids a check-then-act NPE race.

rondagostino commented:
@mwesterby Looks like we have a checkstyle issue:
```
Task :server:checkstyleTest

[ant:checkstyle] [ERROR] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-14942/server/src/test/java/org/apache/kafka/server/AssignmentsManagerTest.java:336:81: '{' is not followed by whitespace. [WhitespaceAround]

[ant:checkstyle] [ERROR] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-14942/server/src/test/java/org/apache/kafka/server/AssignmentsManagerTest.java:336:82: '}' is not preceded with whitespace. [WhitespaceAround]

[ant:checkstyle] [ERROR] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-14942/server/src/test/java/org/apache/kafka/server/AssignmentsManagerTest.java:344:81: '{' is not followed by whitespace. [WhitespaceAround]

[ant:checkstyle] [ERROR] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-14942/server/src/test/java/org/apache/kafka/server/AssignmentsManagerTest.java:344:82: '}' is not preceded with whitespace. [WhitespaceAround]



> Task :server:checkstyleTest FAILED
```

mwesterby commented:
Should be fixed now!

rondagostino commented:
LGTM, failing tests are unrelated.  Merging.

Saboteurs09 commented:
## PR Gatekeeper — Final Review

| **Risk Score** | **Decision** |
|----------------|-------------|
| **18** / 100 | **APPROVE** |

### Summary

This PR is approved for merge from a risk perspective.

### Key risks

- **No AI insights** — AI review was skipped or returned no insights.

### Test gaps

- `server/src/main/java/org/apache/kafka/server/AssignmentsManager.java` — no corresponding test class (suggested: `src/test/java/server/src/main/java/org/apache/kafka/server/AssignmentsManager.java`)

### Test execution

- **Passed:** 0 / 0
- **Time:** 0 ms


---
*This review was generated by the PR Gatekeeper. Please address risks before merging.*

Saboteurs09 reviewed (approved):
## PR Gatekeeper — Final Review

| **Risk Score** | **Decision** |
|----------------|-------------|
| **18** / 100 | **APPROVE** |

### Summary

This PR is approved for merge from a risk perspective.

### Key risks

- **No AI insights** — AI review was skipped or returned no insights.

### Test gaps

- `server/src/main/java/org/apache/kafka/server/AssignmentsManager.java` — no corresponding test class (suggested: `src/test/java/server/src/main/java/org/apache/kafka/server/AssignmentsManager.java`)

### Test execution

- **Passed:** 0 / 0
- **Time:** 0 ms


---
*This review was generated by the PR Gatekeeper. Please address risks before merging.*

Saboteurs09 commented:
## PR Gatekeeper — Final Review

| **Risk Score** | **Decision** |
|----------------|-------------|
| **13** / 100 | **APPROVE** |

### Summary

This PR is approved for merge from a risk perspective.

### Key risks

- **No AI insights** — AI review was skipped or returned no insights.

### Test gaps

- `server/src/main/java/org/apache/kafka/server/AssignmentsManager.java` (suggested: `server/src/test/java/org/apache/kafka/server/AssignmentsManagerTest.java`)

### Test execution

- **Passed:** 0 / 0
- **Time:** 0 ms


---
*This review was generated by the PR Gatekeeper. Please address risks before merging.*

Saboteurs09 reviewed (approved):
## PR Gatekeeper — Final Review

| **Risk Score** | **Decision** |
|----------------|-------------|
| **13** / 100 | **APPROVE** |

### Summary

This PR is approved for merge from a risk perspective.

### Key risks

- **No AI insights** — AI review was skipped or returned no insights.

### Test gaps

- `server/src/main/java/org/apache/kafka/server/AssignmentsManager.java` (suggested: `server/src/test/java/org/apache/kafka/server/AssignmentsManagerTest.java`)

### Test execution

- **Passed:** 0 / 0
- **Time:** 0 ms


---
*This review was generated by the PR Gatekeeper. Please address risks before merging.*

Saboteurs09 commented:
## PR Gatekeeper — Final Review

| **Risk Score** | **Decision** |
|----------------|-------------|
| **13** / 100 | **APPROVE** |

### Summary

This PR is approved for merge from a risk perspective.

### Key risks

- **No AI insights** — AI review was skipped or returned no insights.

### Test execution

- **Passed:** 0 / 0
- **Time:** 0 ms


---
*This review was generated by the PR Gatekeeper. Please address risks before merging.*

Saboteurs09 reviewed (approved):
## PR Gatekeeper — Final Review

| **Risk Score** | **Decision** |
|----------------|-------------|
| **13** / 100 | **APPROVE** |

### Summary

This PR is approved for merge from a risk perspective.

### Key risks

- **No AI insights** — AI review was skipped or returned no insights.

### Test execution

- **Passed:** 0 / 0
- **Time:** 0 ms


---
*This review was generated by the PR Gatekeeper. Please address risks before merging.*

Saboteurs09 commented:
## PR Gatekeeper — Final Review

| **Risk Score** | **Decision** |
|----------------|-------------|
| **13** / 100 | **APPROVE** |

### Summary

This PR is approved for merge from a risk perspective.

### Key risks

- **No AI insights** — AI review was skipped or returned no insights.

### Test execution

- **Passed:** 0 / 0
- **Time:** 0 ms


---
*This review was generated by the PR Gatekeeper. Please address risks before merging.*

Saboteurs09 reviewed (approved):
## PR Gatekeeper — Final Review

| **Risk Score** | **Decision** |
|----------------|-------------|
| **13** / 100 | **APPROVE** |

### Summary

This PR is approved for merge from a risk perspective.

### Key risks

- **No AI insights** — AI review was skipped or returned no insights.

### Test execution

- **Passed:** 0 / 0
- **Time:** 0 ms


---
*This review was generated by the PR Gatekeeper. Please address risks before merging.*
